File: qled.cpp
/***************************************************************************
 *   Copyright (C) 2010 by P. Sereno                                       *
 *   http://www.sereno-online.com                                          *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Lesser General Public License           *
 *   version 2.1 as published by the Free Software Foundation              *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU Lesser General Public License for more details.                   *
 *   http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html.               *
 ***************************************************************************/

#include <QColor>
#include <QtGlobal>
#include <QtGui>
#include <QPolygon>
#include <QtSvg>
#include <QSvgRenderer>

#include "qled.h"

/*!
  \brief QLed: this is the QLed constructor.
  \param parent: The Parent Widget
*/
QLed::QLed(QWidget *parent)
    : QWidget(parent),
      m_value(false),
      m_onColor("Red"),
      m_offColor("Grey"),
      m_shape(Circle),
      renderer(new QSvgRenderer())
{
    shapes << "circle" << "square" << "triang" << "round";
    updateRenderer();
}


QLed::~QLed() {
    delete renderer;
}


/*!
  \brief paintEvent: painting method
  \param QPaintEvent *
  \return void
*/
void QLed::paintEvent(QPaintEvent *)
{
    QPainter painter(this);
    painter.setRenderHint(QPainter::Antialiasing, true);
    renderer->render(&painter);
}


/*!
  \brief updateRenderer: this updates the renderer to contain the correct data for the current properties
  \return void
*/
void QLed::updateRenderer()
{
    QColor currentColor = m_value ? m_onColor : m_offColor;

    QString filePathTemplate(":/resources/%1.svg");
    QString filePath = filePathTemplate.arg(shapes[m_shape]);

    QFile svgFile(filePath);
    svgFile.open(QIODevice::ReadOnly);
    auto svgContent = svgFile.readAll();
    svgFile.close();

    svgContent.replace("{$color_dark}", currentColor.darker().name(QColor::HexRgb).toLatin1());
    svgContent.replace("{$color_light}", currentColor.name(QColor::HexRgb).toLatin1());

    renderer->load(svgContent);
}


/*!
  \brief setOnColor: this method allows to change the On color
  \param QColor newColor
  \return void
*/
void QLed::setOnColor(QColor newColor)
{
    m_onColor = newColor;
    updateRenderer();
    update();
}


/*!
  \brief setOffColor: this method allows to change the Off color
  \param QColor newColor
  \return void
*/
void QLed::setOffColor(QColor newColor)
{
    m_offColor = newColor;
    updateRenderer();
    update();
}


/*!
  \brief setShape: this method allows to change the led shape {Circle,Square,Triangle,Rounded rectangle}
  \param ledColor newColor
  \return void
*/
void QLed::setShape(ledShape newShape)
{
    m_shape = newShape;
    updateRenderer();
    update();
}


/*!
  \brief setValue: this method allows to set the led value {true,false}
  \param ledColor newColor
  \return void
*/
void QLed::setValue(bool value)
{
    m_value = value;
    updateRenderer();
    update();
}


/*!
  \brief toggleValue: this method toggles the led value
  \param ledColor newColor
  \return void
*/
void QLed::toggleValue()
{ 
    m_value =! m_value;
    updateRenderer();
    update();
}

File: Python/qled/qled.py
from PySide6.QtWidgets import QAbstractButton
from PySide6.QtCore import Qt, QByteArray, QFile, QTextStream
from PySide6.QtGui import QColor, QPainter
from PySide6.QtSvg import QSvgRenderer
from . import resources
from enum import Enum


class QLed(QAbstractButton):

    class ledShape(Enum):
        Circle = "circle.svg"
        Square = "square.svg"
        Triangle = "triangle.svg"
        Rounded = "rounded.svg"
        Rectangle = "rect.svg"

    def __init__(self, parent=None):
        super().__init__(parent)

        self._value = False
        self._colorOn = QColor("red")
        self._colorOff = QColor("gray")

        self._renderer = QSvgRenderer()
        self.setShape(self.ledShape.Circle)

        self.setFocusPolicy(Qt.StrongFocus)
        self.setMouseTracking(True)
        self.setAttribute(Qt.WidgetAttribute.WA_MacShowFocusRect, True)

    def paintEvent(self, _event):
        painter = QPainter(self)
        painter.setRenderHint(QPainter.Antialiasing)
        self._renderer.render(painter, self.rect())

    def checkStateSet(self):
        self._updateRenderer()

    def nextCheckState(self):
        super().nextCheckState()
        self._updateRenderer()

    def _updateRenderer(self):
        color = self._colorOn if self._value else self._colorOff
        svgContent = self._shape
        svgContent = svgContent.replace("{$color_dark}", color.darker().name(QColor.HexRgb))
        svgContent = svgContent.replace("{$color_light}", color.lighter().name(QColor.HexRgb))

        self._renderer.load(QByteArray(svgContent.encode("utf-8")))
        self.update()

    def setOnColor(self, color: QColor):
        self._colorOn = color
        self._updateRenderer()

    def setOffColor(self, color: QColor):
        self._colorOff = color
        self._updateRenderer()

    def setShape(self, shape):
        self._shape = self.loadShapeTemplate(shape)
        self._updateRenderer()

    def loadShapeTemplate(self, shapeType):
        shapePath = f":/qled/resources/{shapeType.value}"
        file = QFile(shapePath)
        if not file.open(QFile.ReadOnly | QFile.Text):
            raise FileNotFoundError(f"Cannot open resource: {shapePath}")
        stream = QTextStream(file)
        svgData = stream.readAll()
        file.close()

        return svgData

    def setValue(self, value: bool):
        self._value = value
        self._updateRenderer()

    def toggleValue(self):
        self._value = not self._value
        self._updateRenderer()

    def sizeHint(self):
        return self.minimumSizeHint()

    def minimumSizeHint(self):
        return self._renderer.defaultSize()
